ludwig/encoders/text_encoders.py

```diff
@@ -2374,3 +2380,153 @@ def output_shape(self) -> torch.Size:
+            model_outputs = self.model(input_ids=inputs, output_hidden_states=True).hidden_states[-1]
+
+        return {ENCODER_OUTPUT: model_outputs}
+
+    def _save_to_state_dict(self, destination: Dict, prefix: str, keep_vars: bool):
+        # This is called by `torch.nn.Module.state_dict()` under the hood. `state_dict()` does additional work to
+        # prep the dictionary, get submodule state, and run hooks. Overriding this method only impacts the
+        # contents of the state_dict.
+        # The three args to this method are supplied by Module.state_dict
+        # https://github.com/pytorch/pytorch/blob/8739d1e3f9b08f4282fe79fc8dacd781d16913ff/torch/nn/modules/module.py#L1824
+        if self.config.adapter:
+            # get_peft_model_state_dict geneates a state dict that only contains the adapter weights
+            from peft.utils.save_and_load import get_peft_model_state_dict
+
+            sd = get_peft_model_state_dict(self.model)
+            destination.update(sd)
+
+        else:
+            super()._save_to_state_dict(destination, prefix=prefix, keep_vars=keep_vars)
+
+    def state_dict(self, *args, destination=None, prefix="", keep_vars=False):
+        destination = super().state_dict(destination, prefix=prefix, keep_vars=keep_vars)
+
+        if self.config.adapter:
+            adapter_type_prefix = self.ADAPTER_PARAM_NAME_PREFIX[self.config.adapter.type]
+            exclude_model_keys = [k for k in destination.keys() if adapter_type_prefix not in k]
+
+            for k in exclude_model_keys:
+                del destination[k]
+
+        return destination
+
+    def _load_from_state_dict(
+        self, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+    ):
+        # Call this first to make sure torch can do its usual load. In the adapter case, this should essentially be a
+        # no-op, but the adapter weights will be collected in `unexpected_keys` because PEFT changes the parameter
+        # names under the hood.
+        super()._load_from_state_dict(
+            state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+        )
+
+        if self.config.adapter:
+            # When using an adapter, only the adapter weights are saved, and so we only want to load those weights.
+            # Under the hood, PEFT alters the names of the parameters, which leads to an "unexpected keys" error when
+            # using strict mode. This block uses PEFT's version of `load_state_dict` to handle loading in weights.
+            from peft.utils.save_and_load import set_peft_model_state_dict
+
+            adapter_type_prefix = self.ADAPTER_PARAM_NAME_PREFIX[self.config.adapter.type]
+            peft_model_state_dict = {k: v for k, v in state_dict.items() if adapter_type_prefix in k}
+            set_peft_model_state_dict(self.model, peft_model_state_dict)
+
+            if strict:
```

Comment:
I see, so this code block just removes the peft weights from the unexpected keys dictionary that is being passed around in Torch?

Reply:
Yes. We have to do this because PEFT renames the parameters under the hood. With the normal `.save_pretrained`/`from_pretrained` model checkpointing logic, the conversion seems to be handled behind the scenes. When we do it like this though, torch thinks we just have extra weights, and in strict mode that raises and exception.